### PR TITLE
feat: promisify app.getFileIcon()

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Promisifies `app.getFileIcon` natively.

Todo:
- [x] specs
- [x] docs

/cc @ckerr @deepak1556

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)

#### Release Notes

Notes: Promisifies `app.getFileIcon`

Co-authored-by: Charles Kerr <ckerr@github.com>

### CODE REVIEW

**Comment by codebytere** [comment]
Once this is approved i'll open up a PR to deprecate this in 4.0

File: docs/api/app.md [hunk 1]
```diff
@@ -533,17 +533,16 @@ You can request the following paths by the name:
...
 * `options` Object (optional)
   * `size` String
     * `small` - 16x16
     * `normal` - 32x32
     * `large` - 48x48 on _Linux_, 32x32 on _Windows_, unsupported on _macOS_.
-* `callback` Function
-  * `error` Error
-  * `icon` [NativeImage](native-image.md)
+
+Returns `Promise<icon>` - fulfilled with the app's icon, which is a [NativeImage](native-image.md).
```
Comment (sindresorhus) [inline docs/api/app.md]: `Promise<icon>` looks like a type notation, so it would be more natural to make it `Promise<NativeImage>`.

**Comment by codebytere** [comment]
Some thoughts:

The initial plan for promisification hinged on waiting until Node 8 (which bundled `util.promisify`) and then blanket converting all methods taking callbacks at once. From discussion in that issue, i think (and i agree with this perspective) that native promisification as a goal is more desirable than `util.promisify`, as it provides less of a band-aid over the issue and addresses it at the root. 

However, this means that we can't, then, batch convert over all relevant methods at once, and need to do it in this more piecemeal manner. My proposal would be that we maintain a public project board with our progress on this issue, in addition to an issue in maintainers, and merge these piecemeal rather than bottleneck on each method being converted so that we can flag them all at once.

What are y'all's thoughts?

**Comment by codebytere** [comment]
dangit meant to comment not close and comment

File: atom/browser/api/atom_api_app.cc [hunk 2]
```diff
@@ -1085,16 +1083,16 @@ JumpListResult App::SetJumpList(v8::Local<v8::Value> val,
...
-void App::GetFileIcon(const base::FilePath& path, mate::Arguments* args) {
-  mate::Dictionary options;
-  IconLoader::IconSize icon_size;
-  FileIconCallback callback;
-
+v8::Local<v8::Promise> App::GetFileIcon(const base::FilePath& path,
+                                        mate::Arguments* args) {
   v8::Locker locker(isolate());
-  v8::HandleScope handle_scope(isolate());
+  v8::Isolate::Scope scope(isolate());
```
[the remark this answers is not in the document]
Comment (deepak1556) [reply atom/browser/api/atom_api_app.cc]: We create the promise resolvers as persistent handles `v8::Global<v8::Promise::Resolver> resolver_;`, it shouldn't be affected by the presence of a handle scope.

[the remark this answers is not in the document]
Comment (ckerr) [reply atom/browser/api/atom_api_app.cc]: Could you explain more about the use of `resolver_` here? The previous revision of this PR created the Promise inside the scope of a `HandleScope`, but IIUC objects created in that scope are destroyed when the HandleScope is destroyed -- which means the Promise wouldn't survive past the end of the function and couldn't be used to resolve/reject when the async call finishes?

[the remark this answers is not in the document]
Comment (codebytere) [reply atom/browser/api/atom_api_app.cc]: did we determine whether changes are warranted here?

[the remark this answers is not in the document]
Comment (MarshallOfSound) [reply atom/browser/api/atom_api_app.cc]: Pretty sure this should be a `HandleScope`.  `HandleScope`'s only look after local handles, the promise resolver is a global handle

[the remark this answers is not in the document]
Comment (codebytere) [reply atom/browser/api/atom_api_app.cc]: why might it be crashing with one then, you think?

File: spec/api-app-spec.js [hunk 3]
```diff
@@ -769,61 +769,46 @@ describe('app module', () => {
...
-        app.getFileIcon(iconPath, { size: 'normal' }, (err, icon) => {
-          const size = icon.getSize()
-          expect(err).to.be.null()
-          expect(size.height).to.equal(sizes.normal)
-          expect(size.width).to.equal(sizes.normal)
-          done()
-        })
+      it('fetches a normal icon', async () => {
+        const icon = await app.getFileIcon(iconPath, { size: 'normal' })
+        console.log(icon)
```
Comment (deepak1556) [inline spec/api-app-spec.js]: remove logging

File: atom/browser/api/atom_api_app.cc [hunk 4]
```diff
@@ -1125,16 +1123,16 @@ JumpListResult App::SetJumpList(v8::Local<v8::Value> val,
...
-void App::GetFileIcon(const base::FilePath& path, mate::Arguments* args) {
-  mate::Dictionary options;
-  IconLoader::IconSize icon_size;
-  FileIconCallback callback;
-
+v8::Local<v8::Promise> App::GetFileIcon(const base::FilePath& path,
+                                        mate::Arguments* args) {
   v8::Locker locker(isolate());
-  v8::HandleScope handle_scope(isolate());
+  v8::HandleScope scope(isolate());
```
Comment (deepak1556) [inline atom/browser/api/atom_api_app.cc]: this line can remain unchanged.

**Pull request merged**